src/Storages/MergeTree/TemporaryParts.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <boost/noncopyable.hpp>
+#include <mutex>
+#include <string>
+#include <unordered_set>
+
+namespace DB
+{
+
+/// Manages set of active temporary paths that should not be cleaned by background thread.
+class TemporaryParts : private boost::noncopyable
+{
+private:
+    std::mutex mutex;
+
+    /// NOTE: It is pretty short, so use STL is fine.
+    std::unordered_set<std::string> parts;
+
+public:
+    /// Returns true if passed part name is active.
+    /// (is the destination for one of active mutation/merge).
+    ///
+    /// NOTE: that it accept basename (i.e. dirname), not the path,
+    /// since later requires canonical form.
+    bool contains(const std::string & basename);
```

Comment:
I would consider making the `mutex` object `mutable` and having this function `const`.
@KochetovNicolai not sure if it's aligned with our general practice but I think it's a good thing from API perspective.

Reply:
It could be reasonable, however, if caller side store non-const objects, then I would not care a lot.

Reply:
OK, will add `mutable` (however superfluous `mutable` looks odd to me).

Reply:
I think it makes more sense if you think what is only part of the internals of a class and what is accessible to the user.
`parts` is something the user interacts with, while `mutex` is just a tool to make the class thread-safe, it's not important to the user. So having `contains` const I think is better than avoiding `mutable` in this case.